**Master SQL Development**

_Procedural Language Structure_

In procedural languages (Python, R, C), you specify **how** to do something

\- Directly manipulate data structures in an order determined by code

  

_Execution Plans_

\- SQL query processors take declarative statements

\- They create procedural plans

\- These are the set of steps that scan, filter and join data

  

**Scanning Tables and Indexes**

_Scanning Operations - Linear operation_

\- Scanning looks at each row

\- Fetch data block containing row 

\- Apply filter or condition (based on WHERE)

\- Cost of query based on number of rows in the table

  

Most efficient with small tables, inefficient with large tables. 

  

**_Full table scan_ \-** scans the entire data to retrieve data, inefficient

_**Indexes**_ \-  Ordered subsets of data in a table  

> \- Efficient to look up rows by value
> 
> \- Faster to search index for attribute value
> 
> \- Points to a location of row
> 
> \- Ex: filter by checking index for match, then retrieve row

  

NOTE: SQL tables are NOT ordered. Ordering only happens with the ORDER BY clause, which results in a cursor.

  

**Index Types:** 

\- Balanced Tree, or B-Tree - for equality and Range queries

\- Hash Indexes, for equality

\- Bitmap, for inclusion

\- Specialized Indexes, for geo-spatial or user-defined indexing strategies

**Joining Tables**
How to match rows:
> \- Primary Key in one table   
> \- Foreign Key in the other table
**3 Ways to Join Tables**  
> 1.  Nested Loop Join
> > \- Compare all rows in both tables to each other
> 
> > \- Loop through one table
> 
> > \- For each row, loop through the other table
> 
> > \- At each step, compare keys
> 
> > \- Simple to implement
> 
> > \- Can be expensive
> 
> 2\. Hash Join
> > \- Calculate the hash value of key and join based on matching hash values
> 
> > \- Compute hash values of key values in smaller table
> 
> > \- Store in hash table, which has hash value and row attributes
> 
> > \- Scan larger table; find rows from smaller hash table
> 
> 3\. Sort Merge Join
> > \- Sort both tables and then join rows while taking advantage of order
> 
> > \- Compare rows like nested loop join, but...
> 
> > \- Stop when it is no longer possible to find a match later in the table due to sort order
> 
> > \- Scan the driving table only once
  
**Partitioning Data**
> \- Large tables are stored as smaller tables, known as partitions
> 
> \- Used to improve query, load, and delete operations
> 
> \- Used for large tables
> 
> \- When subset of data is accessed or changed
> 
> \- Can be expensive
  
Need to use a partition key
> \- Based on time... 
> 
> \- Local indexes can be used per partition
> 
> \- Global indexes across all partitions, for data spread across all partitions

**Using EXPLAIN and ANALYZE** 

These give the query plan and the information about the execution of the query

In [2]:
EXPLAIN ANALYZE SELECT * FROM staff


(3 row(s) affected)

Total execution time: 00:00:00.002

QUERY PLAN Seq Scan on staff (cost=0.00..24.00 rows=1000 width=75) (actual time=0.019..0.136 rows=1000 loops=1) Planning Time: 0.066 ms Execution Time: 0.183 ms

In [3]:
EXPLAIN ANALYZE SELECT last_name FROM staff

(3 row(s) affected)

Total execution time: 00:00:00.001

QUERY PLAN Seq Scan on staff (cost=0.00..24.00 rows=1000 width=7) (actual time=0.010..0.086 rows=1000 loops=1) Planning Time: 0.041 ms Execution Time: 0.115 ms

**Query all staff with a salary above 75,000**

In [4]:
SELECT  *
FROM    staff
WHERE   salary > 75000

(717 row(s) affected)

Total execution time: 00:00:00.027

id last_name email gender department start_date salary job_title region_id 3 Carr fcarr2@woothemes.com Male Automotive 2009-07-12 101768 Recruiting Manager 3 4 Murray jmurray3@gov.uk Female Jewelery 2014-12-25 96897 Desktop Support Technician 3 6 Phillips bphillips5@time.com Male Tools 2013-08-21 118497 Executive Secretary 1 8 Harris aharris7@ucoz.com Female Toys 2003-08-12 84427 Safety Technician I 4 9 James rjames8@prnewswire.com Male Jewelery 2005-09-07 108657 Sales Associate 2 10 Sanchez rsanchez9@cloudflare.com Male Movies 2013-03-13 108093 Sales Representative 1 11 Jacobs jjacobsa@sbwire.com Female Jewelery 2003-11-27 121966 Community Outreach Specialist 7 13 Schmidt sschmidtc@state.gov Male Baby 2002-10-13 85227 Compensation Analyst 3 15 Jacobs ajacobse@google.it Female Games 2007-03-04 141139 Community Outreach Specialist 7 16 Medina smedinaf@amazonaws.com Female Baby 2008-03-14 106659 Web Developer III 1 17 Morgan dmorgang@123-reg.co.uk Female Kids 2011-05-04 148952 Programmer IV 6 18 Nguyen jnguyenh@google.com Male Home 2014-11-03 93804 Geologist II 5 19 Day rdayi@chronoengine.com Male Electronics 2004-09-22 109890 VP Sales 3 20 Carr dcarrj@ocn.ne.jp Female Movies 2007-11-22 115274 VP Quality Control 5 21 Bryant sbryantk@wunderground.com Female Industrial 2005-08-12 78052 Software Consultant 3 22 Alexander kalexanderl@marketwatch.com Male Automotive 2011-12-26 144724 Marketing Assistant 2 23 Chapman jchapmanm@archive.org Female Jewelery 2001-04-12 126103 Senior Developer 4 24 Kelly ekellyn@weibo.com Male Garden 2006-06-29 144965 Nurse Practicioner 2 26 Porter vporterp@yelp.com Female Home 2011-05-30 90746 Dental Hygienist 2 27 Owens cowensq@shareasale.com Female Home 2005-05-26 78698 Executive Secretary 2 29 Price cprices@usa.gov Female Clothing 2013-07-05 95436 Quality Engineer 3 30 Weaver dweavert@shinystat.com Male Beauty 2012-02-17 83144 Account Representative III 2 31 Willis hwillisu@army.mil Female Tools 2012-10-21 113507 Accountant I 7 32 Torres ltorresv@amazon.de Male Games 2014-04-07 78245 Nuclear Power Engineer 6 33 Greene wgreenew@blogspot.com Male Movies 2000-03-21 87057 Assistant Manager 1 34 Reed areedx@cisco.com Male Movies 2002-11-28 120579 Assistant Media Planner 6 35 Smith bsmithy@statcounter.com Female Books 2013-06-13 94884 Librarian 7 36 Bradley nbradleyz@goodreads.com Male Outdoors 2008-12-28 107222 Payment Adjustment Coordinator 3 38 Williamson dwilliamson11@rediff.com Female Garden 2012-01-25 82026 Nurse 2 39 Kennedy skennedy12@rediff.com Male Baby 2011-07-02 79339 Civil Engineer 1 40 Hawkins phawkins13@twitpic.com Female Clothing 2011-08-20 76809 Quality Engineer 6 41 Allen jallen14@ustream.tv Female Baby 2001-05-10 89680 Web Developer III 6 42 Watkins mwatkins15@wiley.com Male Computers 2008-04-01 125668 Executive Secretary 7 43 Ferguson mferguson16@qq.com Male Sports 2009-05-29 82960 Account Representative III 6 44 Simmons asimmons17@csmonitor.com Male Movies 2013-12-21 113857 Developer III 2 45 Nguyen mnguyen18@biblegateway.com Male Automotive 2002-01-24 108378 Electrical Engineer 7 46 Lawrence jlawrence19@linkedin.com Male Outdoors 2011-12-11 133424 Junior Executive 7 48 Johnson jjohnson1b@usgs.gov Male Computers 2004-04-26 134940 Financial Advisor 4 49 Crawford hcrawford1c@dell.com Male Home 2008-12-21 119471 Editor 6 50 Armstrong parmstrong1d@hc360.com Female Automotive 2010-03-16 89969 Senior Editor 4 51 Dixon jdixon1e@dot.gov Male Baby 2006-09-26 106281 Senior Developer 3 52 Snyder ssnyder1f@ca.gov Female Kids 2002-09-26 141678 Design Engineer 3 53 Coleman dcoleman1g@dedecms.com Male Computers 2008-06-14 110839 Automation Specialist I 1 54 Reed sreed1h@intel.com Female Toys 2001-04-15 84275 VP Marketing 5 56 Watkins kwatkins1j@youtu.be Male Garden 2002-12-02 93680 VP Accounting 3 60 Lane elane1n@un.org Male Outdoors 2004-07-15 104896 Financial Advisor 5 61 Watson cwatson1o@paypal.com Female Tools 2009-08-10 81870 Occupational Therapist 7 63 Young hyoung1q@cisco.com Female Grocery 2007

In [5]:
-- Note: the number of rows is actually off by 2, but this is just an estimate
EXPLAIN 
SELECT  *
FROM    staff
WHERE   salary > 75000

(2 row(s) affected)

Total execution time: 00:00:00.001

QUERY PLAN Seq Scan on staff (cost=0.00..26.50 rows=715 width=75) Filter: (salary > 75000)

In [6]:
EXPLAIN 
ANALYZE
SELECT  *
FROM    staff
WHERE   salary > 75000

(5 row(s) affected)

Total execution time: 00:00:00.002

QUERY PLAN Seq Scan on staff (cost=0.00..26.50 rows=715 width=75) (actual time=0.014..0.102 rows=717 loops=1) Filter: (salary > 75000) Rows Removed by Filter: 283 Planning Time: 0.059 ms Execution Time: 0.129 ms

**Using Indexes to reduce Query Time**

In [7]:
CREATE INDEX idx_staff_salary ON staff(salary)

Commands completed successfully

Total execution time: 00:00:00.016

In [11]:
EXPLAIN ANALYZE SELECT * FROM staff WHERE salary > 75000

(5 row(s) affected)

Total execution time: 00:00:00.002

QUERY PLAN Seq Scan on staff (cost=0.00..26.50 rows=715 width=75) (actual time=0.015..0.131 rows=717 loops=1) Filter: (salary > 75000) Rows Removed by Filter: 283 Planning Time: 0.093 ms Execution Time: 0.161 ms

In [12]:
EXPLAIN ANALYZE SELECT * FROM staff WHERE salary > 150000

(4 row(s) affected)

Total execution time: 00:00:00.001

QUERY PLAN Index Scan using idx_staff_salary on staff (cost=0.28..8.29 rows=1 width=75) (actual time=0.002..0.003 rows=0 loops=1) Index Cond: (salary > 150000) Planning Time: 0.097 ms Execution Time: 0.012 ms

Note: the first scan determined it was faster to just scan everything without using the index, 717 out of 1000.

When filtering for salaries above 150000, it used the indexes prior to filtering, returning only 1 row out of 1000.

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Types of Indexes**

Purpose of Indexes

> \- Speed up access to data  
> \- Help enforce constraints
> 
> \- Indexes are ordered
> 
> \- Typically smaller than tables

  

Reading Data: 

> \- From Memory:    100 ns

> \- From SSD:           1,000,000 ns   (1 ms)
> 
> \- From HDD:          20,000,000 ns (20 ms)

  

Implementing Indexes: 

> \- Data structure is separate from table
> 
> \- Sometimes duplicates some data, for example, keys
> 
> \- Organized differently from tables (ordered, whereas tables are not ordered)

  

While these do add duplication to the existing data, speeds up the data processing

  

**\- B-Trees**  
**\- Bitmap**  
**\- Hash**  

**\- Special Purpose**

**\-------------------------------------------------**

**B-Tree Indexes (Balanced Tree)**

> \- Splits values that are greater than base node to the right, less than base node to the left
> 
> \- Continues this for each node value until all values are split - one node per row

> \- Most common type of index

> \- Used when a large number of possible values in a column (high cardinality)

> \- Rebalances as needed to keep the sides of the tree/tables balanced

> \- Time to access is based on the depth of the tree (log of number of nodes in the tree)

In [13]:
SELECT  * 
FROM    staff
WHERE   email = 'bphillips5@time.com'

(1 row(s) affected)

Total execution time: 00:00:00.001

id last_name email gender department start_date salary job_title region_id 6 Phillips bphillips5@time.com Male Tools 2013-08-21 118497 Executive Secretary 1

In [14]:
-- Used 26.5 computational units
EXPLAIN
SELECT  * 
FROM    staff
WHERE   email = 'bphillips5@time.com'

(2 row(s) affected)

Total execution time: 00:00:00.001

QUERY PLAN Seq Scan on staff (cost=0.00..26.50 rows=1 width=75) Filter: ((email)::text = 'bphillips5@time.com'::text)

In [17]:
-- Used 8.29 computational units with the index

-- B-Tree is the default index in postgres, so no need to specify index type
CREATE INDEX idx_staff_email ON staff(email);

EXPLAIN
SELECT  * 
FROM    staff
WHERE   email = 'bphillips5@time.com'

(2 row(s) affected)

Total execution time: 00:00:00

QUERY PLAN Index Scan using idx_staff_email on staff (cost=0.28..8.29 rows=1 width=75) Index Cond: ((email)::text = 'bphillips5@time.com'::text)

**\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_**

**Bitmap Index**

Good for low cardinality problems

  

> \- Uses a number of bits for each distinct value in a column : yes/no = 2 bits,  Ranks from 1-5 = 5 bits

> \- Can quickly perform Boolean Operations

> \- Used when small number of possible values in a column (low cardinality)

> \- Filter by bitwise operations, such as AND, OR, NOT

> \- Time to access is based on time to perform bitwise operations (longer)

> \- Read-intensive use cases, few writes

>   

  

> \- Some dbs allow you to create bitmap indexes explicitly
> 
> \- Postgres does not
> 
> \- postgres builds bitmap indexes on the fly as needed